### PR TITLE
Make telegram.Bot comparable

### LINKED ISSUE #2319: [FEATURE] Make tg.Bot comparable
`tg.Bot` didn't get rich comparison in #1724, though there are use cases when comparison of `Bot` with `User` objects would be helpful. Most notable when checking if the bot is one of the new chat members reportet in a status update; with rich comparison that would be as easy as

```python
context.bot in update.effective_message.new_chat_members
```

Currently one can use `bot_instance.bot`, which holds the `User` instance for comparison, but it's undocumented and unintuitive. Proposal:

* Make `Bot.__eq__(self, other)` basically return `self.bot == other` and document the behaviour accordingly
* we need to make sure that `self.bot` is already set. For the `Bot.{id, username, …}` properties, that's currently done with the `@info` decorator. I think it would be cleaner to just convert `self.bot` into a property that calls `self.get_me()` if needed.

### PR DESCRIPTION
**Pull request by starry69**
Closes #2319

### CODE REVIEW

File: telegram/bot.py [hunk 1]
```diff
@@ -303,67 +287,68 @@ def _validate_token(token: str) -> str:
         return token
 
     @property  # type: ignore
-    @info
+    def bot(self) -> User:
+        """:class:`telegram.User`: Instance for the bot."""
```
Comment (Bibo-Joshi) [inline telegram/bot.py:291]:
```suggestion
        """:class:`telegram.User`: User instance for the bot as returned by :meth:`get_me`."""
```

File: telegram/bot.py [hunk 2]
```diff
@@ -303,67 +287,68 @@ def _validate_token(token: str) -> str:
...
+
+        if self._bot is None:
+            self.get_me()
```
Comment (Bibo-Joshi) [inline telegram/bot.py:294]: this is probably why everthing's  failing ;)

```suggestion
            self._bot = self.get_me()
```

File: telegram/bot.py [hunk 3]
```diff
@@ -303,67 +287,68 @@ def _validate_token(token: str) -> str:
...
         return self.bot.supports_inline_queries  # type: ignore
 
     @property  # type: ignore
-    @info
     def commands(self) -> List[BotCommand]:
         """List[:class:`BotCommand`]: Bot's commands."""
 
-        return self._commands or []
+        if self._commands is None:
+            self.get_my_commands()
```
Comment (Bibo-Joshi) [inline telegram/bot.py:350]:
```suggestion
           self._commands = self.get_my_commands()
```

File: tests/test_bot.py [hunk 4]
```diff
@@ -145,6 +145,12 @@ def test_get_me_and_properties(self, bot):
         assert f'https://t.me/{get_me_bot.username}' == bot.link
         assert commands == bot.commands
 
+    def test_comparison(self, bot):
```
Comment (Bibo-Joshi) [inline tests/test_bot.py:148]: I think this test should be elaborated. please have a look at the other TG object classes tests for inspiration ;)

**Review by Bibo-Joshi — CHANGES_REQUESTED** [review]

File: telegram/bot.py (hunk 2 above)
Comment (starry69) [reply telegram/bot.py:294]: Oh i thought as `bot.get_me()` call sets attr anyway, so we can leave that...

**Review by Bibo-Joshi — CHANGES_REQUESTED** [review]
Mh, actually you're right about `self.get_me()` already setting `self._bot`, but imo it's still more readible to assign the return value anyway.

Some more notes:

* Please document that bot is now comparable (See e.g. [here](https://github.com/python-telegram-bot/python-telegram-bot/blob/be54cf4ecedfd52c1aa684eb03ad5ef9206b4b1a/telegram/chat.py#L62)) probably just do it als
  ```python
  .. versionadded:: 13.2
     Objects of this class …
  ```
* After a closer look the tests fail b/c some tests want to monkeypatch `bot.bot`, which is no longer possible. Ig just making them patch `bot._bot` instead should be fine
* Some other tests fail, b/c `Bot` is no longer hashable. `Bot.__hash__` should return `hash(self.bot)`, as that's what we're comparing with. see [here](https://docs.python.org/3.5/reference/datamodel.html#object.__hash__)
* For what the equality tests should look like please see [here](https://github.com/python-telegram-bot/python-telegram-bot/blob/be54cf4ecedfd52c1aa684eb03ad5ef9206b4b1a/tests/test_user.py#L539)

File: telegram/bot.py [hunk 5]
```diff
@@ -144,6 +128,10 @@ def decorator(self: 'Bot', *args: Any, **kwargs: Any) -> RT:  # pylint: disable=
 class Bot(TelegramObject):
     """This object represents a Telegram Bot.
 
+    .. versionadded:: 13.2
+    Objects of this class are comparable in terms of equality. Two objects of this class are
+    considered equal, if their :attr:`bot` is equal.
+
```
Comment (Bibo-Joshi) [inline telegram/bot.py:134]: Can you build the docs locally and check if this is rendered correctly? IIRC the text below the directive needs to be indented for sphinx to see it as explanation.

File: telegram/bot.py [hunk 6]
```diff
@@ -302,68 +290,69 @@ def _validate_token(token: str) -> str:
...
-        return self.bot.first_name  # type: ignore
+        return self.bot.first_name
 
-    @property  # type: ignore
-    @info
+    @property
     def last_name(self) -> str:
         """:obj:`str`: Optional. Bot's last name."""
 
         return self.bot.last_name  # type: ignore
```
Comment (Bibo-Joshi) [inline telegram/bot.py:317]: Do we still need those `# type: ignore` comments? 🤔 up to now that fact that `self.bot` was not `None` was hidden in the `@info` decorator, so I guess now mypy can do without them

**Review by Bibo-Joshi — COMMENTED** [review]
Thanks for the updates! Just two comments left :)

File: telegram/bot.py (hunk 6 above)
Comment (starry69) [reply telegram/bot.py:317]: IMG_20210116_234042_007
Ig it's looks fine 🤔

File: telegram/bot.py (hunk 5 above)
Comment (starry69) [reply telegram/bot.py:134]: IMG_20210116_234042_007
Ig it looks fine 🤔

File: telegram/bot.py (hunk 6 above)
Comment (starry69) [reply telegram/bot.py:317]: Umm.. i removed most `# type: ignore` comments but removing those 2 was causing mypy to fail in my local environment. So i left them as it was

File: telegram/bot.py (hunk 5 above)
Comment (Bibo-Joshi) [reply telegram/bot.py:134]: mh, it's supposed to look like this:

image

File: telegram/bot.py (hunk 6 above)
Comment (Bibo-Joshi) [reply telegram/bot.py:317]: okay, let's leave them for now

File: telegram/bot.py (hunk 5 above)
Comment (starry69) [reply telegram/bot.py]: Oops, okay

**Pull request merged**